### PR TITLE
Add timestamps to actions protocol binding - closes #101

### LINKED ISSUE #101: Actions need a way to express time stamps
See #100

### PR DESCRIPTION
**Pull request by benfrancis**
Closes #101

***
<a href="https://pr-preview.s3.amazonaws.com/benfrancis/wot-profile/pull/127.html" title="Last updated on Nov 25, 2021, 9:52 AM UTC (f8ce598)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-profile/127/d51d816...benfrancis:f8ce598.html" title="Last updated on Nov 25, 2021, 9:52 AM UTC (f8ce598)">Diff</a>

***
<a href="https://pr-preview.s3.amazonaws.com/benfrancis/wot-profile/pull/127.html" title="Last updated on Apr 13, 2022, 4:38 PM UTC (d352bc2)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-profile/127/d51d816...benfrancis:d352bc2.html" title="Last updated on Apr 13, 2022, 4:38 PM UTC (d352bc2)">Diff</a>

### CODE REVIEW

**Comment by mlagally** [comment]
@benfrancis 
This looks good, a few comments:
We have to change our mindset wrt. optionality for the profile. Whereas the TD itself makes almost everything optional to enable description of brown field devices, the profile should contain reasonable normative guarantees.
Therefore I suggest to make these timestamps mandatory.

**Comment by mlagally** [comment]
One more thing - Suggest to combine timeFailed and timeSucceeded into a single timestamp: name could be something like timeFinished, timeTerminated, ....

**Comment by benfrancis** [comment]
> I suggest to make these timestamps mandatory.
...
> Suggest to combine timeFailed and timeSucceeded into a single timestamp: name could be something like timeFinished, timeTerminated

@mlagally Please see the comments I made in #101 and https://github.com/w3c/wot-profile/issues/101#issuecomment-966217434 regarding both of these points last week.

> We have to change our mindset wrt. optionality for the profile. Whereas the TD itself makes almost everything optional to enable description of brown field devices, the profile should contain reasonable normative guarantees.

Whilst I agree that Profiles should be more prescriptive than the Thing Description specification in general, in many cases I think it's sufficient to say that Consumers MUST support a feature and Producers MAY use it.

**Comment by mlagally** [comment]
> Whilst I agree that Profiles should be more prescriptive than the Thing Description specification in general, in many cases I think it's sufficient to say that Consumers MUST support a feature and Producers MAY use it.

I think this statement is problematic and may hinder adoption.
The profile is a contract between consumers and things and should minimize implementation complexity on both components.

File: index.html [hunk 1]
```diff
@@ -2021,6 +2021,33 @@ <h6 id="ActionStatus"><code>ActionStatus</code> object</h6>
...
+                  A timestamp indicating the time at which the Thing finished 
+                  executing the action, in ISO format [[ISO8601-1]].
+                </td>
+                <td>optional</td>
+                <td><code>string</code></td>
+              </tr>
+              <tr class="rfc2119-table-assertion">
+                <td><code>timeFailed</code></td>
+                <td>
+                  A timestamp indicating the time at which the Action failed to
```
Comment (mlagally) [inline index.html:2045]: Please consider to eliminate timeFailed. 
This information should be conveyed in the action response.

**Review by mlagally — CHANGES_REQUESTED** [review]

**Comment by benfrancis** [comment]
Regarding the feedback here and in #101...

One of the reasons I originally made the timestamps optional was because you said that some devices may not be able to generate timestamps because they don't have a clock. I therefore also did the same for events.

However, I have now:
1. Combined `timeCompleted` and `timeFailed` into `timeEnded` (`timeFinished` sounds too much like completed and `timeTerminated` sounds too much like it failed)
2. Made `timeRequested` mandatory

A couple of additional points:
1. I considered renaming `timeRequested` to `timeStarted` to better match `timeEnded` but then I realised that changes the meaning. There's an ambiguity between time requested and time started, which may not be the same if the Thing uses an action queue. That is may be a separate issue.
2. I was going to make `timeEnded` mandatory too but that doesn't really make sense for an action which hasn't yet ended!

**Review by mlagally — APPROVED** [review]

File: index.html [hunk 2]
```diff
@@ -2021,6 +2021,25 @@ <h6 id="ActionStatus"><code>ActionStatus</code> object</h6>
                 <td>optional</td>
                 <td><code>string</code></td>
               </tr>
+              <tr class="rfc2119-table-assertion">
+                <td><code>timeRequested</code></td>
+                <td>
+                  A timestamp indicating the time at which the Thing received 
+                  the request to execute the action, in ISO format [[ISO8601-1]].
```
[the remark this answers is not in the document]
Comment (benfrancis) [reply index.html]: I've added a reference to the additional constraints in the datetime section.

File: index.html [hunk 3]
```diff
@@ -2021,6 +2021,25 @@ <h6 id="ActionStatus"><code>ActionStatus</code> object</h6>
...
+                </td>
+                <td>mandatory</td>
+                <td><code>string</code></td>
+              </tr>
+              <tr class="rfc2119-table-assertion">
+                <td><code>timeEnded</code></td>
+                <td>
+                  A timestamp indicating the time at which the Thing
+                  successfully completed executing the action, or failed to
```
[the remark this answers is not in the document]
Comment (benfrancis) [reply index.html:2039]: The `status` member would be set to `failed`.

**Review by mlagally — APPROVED** [review]

**Comment by mlagally** [comment]
Profile call on March 30th:

There are devices with no RTC, it is hard to have a time value on them.
Consider to define a "duration" instead.
Precision of the value is also pontially an interoperability issue.
Need to more carefully consider use cases - is a sequence number and elapsed time sufficient for timeouts?

**Comment by benfrancis** [comment]
> My suggestion:
> 
> * Revert `timeRequested` to being optional
> * Provide implementation guidance to say that Consumers should not assume that the clock of a Thing is set correctly and should therefore `timeEnded` as being relative to `timeRequested`, but not necessarily relative to the clock of the Consumer or the clocks of other Things or other Consumers.

I've pushed a commit which implements these changes. Please let me know what you think.

**Pull request merged**